vowpalwabbit/io_buf.h

```diff
@@ -85,88 +100,46 @@ class io_buf
     return _hash;
   }
 
-  virtual int open_file(const char* name, bool stdin_off) { return open_file(name, stdin_off, READ); }
-
-  virtual int open_file(const char* name, bool stdin_off, int flag = READ)
+  void add_file(std::unique_ptr<VW::io::reader>&& file)
   {
-    int ret = -1;
-    switch (flag)
-    {
-      case READ:
-        if (*name != '\0')
-        {
-#ifdef _WIN32
-          // _O_SEQUENTIAL hints to OS that we'll be reading sequentially, so cache aggressively.
-          _sopen_s(&ret, name, _O_RDONLY | _O_BINARY | _O_SEQUENTIAL, _SH_DENYWR, 0);
-#else
-          ret = open(name, O_RDONLY | O_LARGEFILE);
-#endif
-        }
-        else if (!stdin_off)
-#ifdef _WIN32
-          ret = _fileno(stdin);
-#else
-          ret = fileno(stdin);
-#endif
-        if (ret != -1)
-          files.push_back(ret);
-        break;
-
-      case WRITE:
-#ifdef _WIN32
-        _sopen_s(&ret, name, _O_CREAT | _O_WRONLY | _O_BINARY | _O_TRUNC, _SH_DENYWR, _S_IREAD | _S_IWRITE);
-#else
-        ret = open(name, O_CREAT | O_WRONLY | O_LARGEFILE | O_TRUNC, 0666);
-#endif
-        if (ret != -1)
-          files.push_back(ret);
-        break;
+    assert(output_files.size() == 0);
+    input_files.push_back(std::move(file));
+  }
 
-      default:
-        std::cerr << "Unknown file operation. Something other than READ/WRITE specified" << std::endl;
-        ret = -1;
-    }
-    if (ret == -1 && *name != '\0')
-      THROWERRNO("can't open: " << name);
-    return ret;
+  void add_file(std::unique_ptr<VW::io::writer>&& file)
+  {
+    assert(input_files.size() == 0);
+    output_files.push_back(std::move(file));
   }
 
-  virtual void reset_file(int f)
+  void reset_file(VW::io::reader* f)
   {
-#ifdef _WIN32
-    _lseek(f, 0, SEEK_SET);
-#else
-    lseek(f, 0, SEEK_SET);
-#endif
+    f->reset();
```

Comment:
should check is_resettable() first?  if not is_resettable() close and open?

Reply:
Yeah checking first makes sense. But if it is not resettable then there's nothing you can do, so you'll either propagate the error or swallow it. You can't close and reopen because you don't know what it is.